src/csexp.ml

```diff
@@ -29,107 +37,135 @@ module Make (Sexp : Sexp) = struct
     let int_of_digit c = Char.code c - Char.code '0'
 
     let rec parse_atom input len =
-      match Input.read_char input with
-      | '0' .. '9' as c ->
+      Input.Monad.bind (Input.read_char input) @@ function
+      | Result.Ok c when '0' <= c && c <= '9' ->
         let len = (len * 10) + int_of_digit c in
         if len > Sys.max_string_length then
-          parse_error "atom too big to represent"
+          Input.Monad.return @@ parse_error "atom too big to represent"
         else
           parse_atom input len
-      | ':' ->
-        let s = Input.read_string input len in
-        Atom s
-      | _ -> invalid_character ()
-
-    let rec parse_many input depth acc =
-      match Input.read_char input with
-      | '(' ->
-        let sexps = parse_many input (depth + 1) [] in
-        parse_many input (depth + 1) (List sexps :: acc)
-      | ')' ->
+      | Result.Ok ':' -> (
+        Input.Monad.bind (Input.read_string input len) @@ function
+        | Result.Ok s -> Input.Monad.return @@ Result.Ok (Atom s)
+        | Result.Error e -> Input.Monad.return @@ Result.Error e )
+      | Result.Ok c -> Input.Monad.return @@ invalid_character c
```

Comment:
In this code we are using a single monad, so we can just open `Input.Monad` or re-bind `return`. The code will be as clear and a bit less verbose. Same for `Ok` and `Error`, there is no ambiguity as to what they mean here. Let's rely on type-based disambiguation to drop the `Result.`